apps/emqx/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,322 @@
+    transaction(fun del_stale_mfa/0),
+    {keep_state_and_data, [catch_up_after(10 * 60)]};
+
+handle_event(info, {mnesia_table_event, {write, #cluster_rpc_mfa{} = MFARec, _AId}}, ?IDLE, Node) ->
+    handle_mfa_write_event(MFARec, Node);
+handle_event(info, {mnesia_table_event, {write, _MFARec, _ActivityId}}, ?CATCH_UP, _Node) ->
+    {keep_state_and_data, [catch_up_after(1)]};
+
+handle_event({call, From}, reset, _State, _Node) ->
+    _ = ekka_mnesia:clear_table(?CLUSTER_CURSOR),
+    _ = ekka_mnesia:clear_table(?CLUSTER_MFA),
+    {keep_state_and_data, [{reply, From, ok}, catch_up_after(1)]};
+
+handle_event({call, From}, {commit, MFA}, ?IDLE, Node) ->
+    case transaction(fun() -> init_mfa(Node, MFA) end) of
+        {atomic, {ok, TnxId}} ->
+            {keep_state, Node, [{reply, From, {ok, TnxId}}, del_mfa_after(5 * 60)]};
+        {aborted, Reason} ->
+            {keep_state, Node, [{reply, From, {error, Reason}}, del_mfa_after(5 * 60)]}
+    end;
+handle_event({call, From}, {commit, _MFA}, ?CATCH_UP, Node) ->
+    case catch_up(Node) of
+        {next_state, ?IDLE, Node, _Actions} ->
+            {next_state, ?IDLE, Node, [{postpone, true}]};
+        _ ->
+            Reason = "There are still transactions that have not been executed.",
+            {keep_state, Node, [{reply, From, {error, Reason}}, catch_up_after(1)]}
+    end;
+
+handle_event(_EventType, _EventContent, _StateName, _Node) ->
+    {keep_state_and_data, [del_mfa_after(5 + rand:uniform(5))]}.
+
+terminate(_Reason, _StateName, _Node) ->
+    ok.
+
+code_change(_OldVsn, StateName, Node, _Extra) ->
+    {ok, StateName, Node}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+catch_up(Node) ->
+    case get_next_mfa(Node) of
+        {atomic, caught_up} -> {next_state, ?IDLE, Node, [del_mfa_after(5 * 60)]};
+        {atomic, {still_lagging, NextId, MFA}} ->
+            case apply_mfa(NextId, MFA) of
+                ok ->
+                    Fun = fun() ->
+                        Rec = #cluster_rpc_cursor{node = Node, tnx_id = NextId},
+                        mnesia:write(?CLUSTER_CURSOR, Rec, write)
+                          end,
+                    case transaction(Fun) of
+                        {atomic, ok} -> catch_up(Node);
+                        _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+                    end;
+                _ -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+            end;
+        {aborted, _Reason} -> {next_state, ?CATCH_UP, Node, [catch_up_after(1)]}
+    end.
+
+get_next_mfa(Node) ->
+    Fun =
+        fun() ->
+            case mnesia:read(?CLUSTER_CURSOR, Node, write) of
+                [] ->
+                    LatestId = get_latest_id(),
+                    Cursor = #cluster_rpc_cursor{node = Node, tnx_id = max(LatestId - 1, 0)},
```

Comment:
Yes, When both MFA and COMMIT tables are empty (initial state), each node will catch up and then set its tnx_id to 0 instead of -1.

Reply:
We want `max(LatestId-1,0)` not `max(LatestId,0)`, because:

The initial state updates the MFA, if other nodes do catch up before receiving the EVENT, here, it will write its state as the latest MFA, but he himself has not executed this MFA, so it must start with latestId-1.

The same logic as when it receives an MFA event and can't find its own progress, it sets it to EventID-1.